### PR TITLE
Refactoring: make 'struct hash' behave as a value semantic object

### PR DESCRIPTION
**Pull request by AlexanderLanin**
Currently hash is allocated on heap which makes it necessary to handle it as a pointer and to remember to free it. While this commit is obviously not the final achievable result it seems trivial to integrate it while already achieving fewer memory leaks (e.g. cpp_hash in ccache.cpp)

### CODE REVIEW

File: src/hash.cpp [hunk 1]
```diff
@@ -63,32 +64,37 @@ do_debug_text(struct hash* hash, const void* s, size_t len)
...
 {
-  free(hash);
+  pimpl->state = other.state;
+  pimpl->debug_binary = NULL;
+  pimpl->debug_text = NULL;
 }
 
+
+// Force destructor to cpp file.
+// Required so unique_ptr can call delete on a known object.
```
Comment (jrosdahl) [inline src/hash.cpp]:
```suggestion
// The destructor needs to be in the cpp file for std::unique_ptr to see
// the complete class definition.
```

File: src/hash.hpp [hunk 2]
```diff
@@ -42,14 +44,25 @@ bool digests_equal(const struct digest* d1, const struct digest* d2);
 // struct hash represents the hash algorithm's inner state.
 struct hash;
 
-// Create a new hash state.
-struct hash* hash_init(void);
+// value semantic hash to replace 'struct hash' in the future
+class Hash {
```
Comment (jrosdahl) [inline src/hash.hpp]: How about implementing the class in `Hash.hpp`/`Hash.cpp` instead?

Comment (AlexanderLanin) [reply src/hash.hpp]: You are technically quite right, but as not everyone has a filesystem which is case sensitive it feels like this should be avoided. I would therefore rename the files as a last step.

Comment (jrosdahl) [reply src/hash.hpp]: Right, I solved that problem for utils by renaming utils to legacy_utils. But OK.

**Comment by jrosdahl** [comment]
@AlexanderLanin: Would you like to rebase this PR?

**Comment by AlexanderLanin** [comment]
Merging this old PR is more effort then restarting from scratch. Closing PR.

**Pull request closed without merging**